### PR TITLE
feat(core): findAllApplications order by createdAt desc

### PR DESCRIPTION
**Pull request by IceHe**
## Summary

- feat(core): find many data order by
- feat(core): findAllApplications order by createdAt desc

## Linear Issue Reference

LOG-1855

## Testing

image
image

---
image

---
@logto-io/eng

### CODE REVIEW

File: packages/core/src/database/types.ts [hunk 1]
```diff
@@ -6,10 +6,15 @@ export type FieldIdentifiers<Key extends string | number | symbol> = {
   [key in Key]: IdentifierSqlTokenType;
 };
 
+export type Order = 'asc' | 'desc';
```
Comment (wangsijie) [inline packages/core/src/database/types.ts:9]: how about change the name to `OrderDirection` to avoid confusing.

**Review by wangsijie — APPROVED** [review]

**Comment by IceHe** [comment]
Implement it like `connectKeyValueWithEqualSign`. See #347 .

From

```ts
  const buildOrderBy = (data: OrderBy<Schema>) =>
    Object.entries(data)
      .filter((entry): entry is [keyof Schema & string, Order] => {
        const [key, order] = entry;

        return isKeyOfSchema(key) && order !== undefined;
      })
      .map(([key, order]) =>
        order === 'asc' ? sql`${fields[key]} asc` : sql`${fields[key]} desc`
      );
```

to

```ts
  const buildOrderBy = (data: OrderBy<Schema>) =>
    Object.entries(data)
      .map(
        ([key, order]) =>
          isKeyOfSchema(key) &&
          order !== undefined &&
          (order === 'asc' ? sql`${fields[key]} asc` : sql`${fields[key]} desc`)
      )
      .filter((value): value is Truthy<typeof value> => notFalsy(value));
```

File: packages/core/src/database/find-many.ts [hunk 2]
```diff
@@ -21,14 +21,28 @@ export const buildFindMany = <Schema extends SchemaLike, ReturnType extends Sche
...
       .filter((value): value is Truthy<typeof value> => notFalsy(value));
 
-  return async ({ where, limit, offset }: FindManyData<Schema> = {}) => {
+  const buildOrderBy = (data: OrderBy<Schema>) =>
+    Object.entries(data)
+      .map(
+        ([key, order]) =>
+          isKeyOfSchema(key) &&
+          order !== undefined &&
+          (order === 'asc' ? sql`${fields[key]} asc` : sql`${fields[key]} desc`)
```
Comment (IceHe) [inline packages/core/src/database/find-many.ts:30]: sql.raw

**Review by darcyYe — APPROVED** [review]

File: packages/core/src/database/find-many.ts [hunk 3]
```diff
@@ -21,14 +21,27 @@ export const buildFindMany = <Schema extends SchemaLike, ReturnType extends Sche
       )
       .filter((value): value is Truthy<typeof value> => notFalsy(value));
 
-  return async ({ where, limit, offset }: FindManyData<Schema> = {}) => {
+  const buildOrderBy = (data: OrderBy<Schema>) =>
```
Comment (xiaoyijun) [inline packages/core/src/database/find-many.ts:24]: This logic seems the same as `connectKeyValueWithEqualSign`.
Add an issue to refactor this?

Comment (IceHe) [reply packages/core/src/database/find-many.ts:24]: The data type is not the same.
(`Partial<Schema>` and `OrderBy<Schema>`)
If extracted, the implementation is a little complex.

Comment (xiaoyijun) [reply packages/core/src/database/find-many.ts]: so leave an issue for this?

Comment (xiaoyijun) [reply packages/core/src/database/find-many.ts]: It's not a blocker.

**Review by xiaoyijun — APPROVED** [review]
LGTM

File: packages/core/src/database/find-many.ts (hunk 3 above)
Comment (IceHe) [reply packages/core/src/database/find-many.ts]: ok

**Pull request merged**